### PR TITLE
Wan Control Trainer

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
This PR adds support for channel-concatenated control condition training for Wan, and it is part of #310.

With video models, you get another dimension of controllability. Recently, we've seen a variety of applications that support interpolating between images, adding effects, advanced controllability by extra input signals, video extension, image conditioning, etc. The goal with this PR is to enable all of that with this simple architecture modification. See #310 for more details.

This PR introduces `FrameConditioningType`, which is specifically used only for video models. The following conditioning types are currently supported:
- INDEX: Condition on any arbitrary latent frame. The only sensible choice is to set the `--frame_conditioning_index` to `0`, but it is simple to implement for any frame and enables your wacky lora ideas to come true!
- PREFIX: Condition on some number of start frames. This is most similar to NVIDIA's Cosmos Video2World model, where you can extend an image/video
- RANDOM: Condition on randomly selected frames. This technique allows multiple images to be used for interpolation. Give your crazy lora ideas a life!
- FIRST_AND_LAST: Condition only on first and last frames for interpolating between two images.
- FULL: This is the default setting. It's best used when you want to condition on an entire different video for generation. For example, providing an input video for editing, or canny/depth/pose control.

Additionally, the option `--frame_conditioning_concatenate_mask` can be used to add a single channel frame mask indicating to the model what frames are used for conditioning. It has not been rigourously tested, but may help with model convergence. More experiments need to be carried out to understand its effectiveness.

### Inference

The following results are from a 10k step run on a single GPU with first frame conditioning. The other runs with different conditioning types also succeeded and seem to be working well. However, 10k steps with batch_size=1 is too low for convergence and the model does not generalize well at this scale. Bigger runs are needed for releasable checkpoints, and will be queued after the trainer is made faster!

```python
import numpy as np
import torch
from diffusers import WanPipeline
from diffusers.utils import export_to_video, load_video
from finetrainers.trainer.control_trainer.data import apply_frame_conditioning_on_latents
from finetrainers.models.utils import _expand_conv3d_with_zeroed_weights
from finetrainers.patches import load_lora_weights
from finetrainers.patches.dependencies.diffusers.control import control_channel_concat

dtype = torch.bfloat16
device = torch.device("cuda")
generator = torch.Generator().manual_seed(0)

pipe = WanPipeline.from_pretrained("Wan-AI/Wan2.1-T2V-1.3B-Diffusers", torch_dtype=dtype).to(device)

in_channels = pipe.transformer.config.in_channels
patch_channels = pipe.transformer.patch_embedding.in_channels
pipe.transformer.patch_embedding = _expand_conv3d_with_zeroed_weights(pipe.transformer.patch_embedding, new_in_channels=2 * patch_channels)

load_lora_weights(pipe, "/raid/aryan/wan-control-image-condition", "wan-lora")
pipe.to(device)

prompt = "The video shows a vibrant green Mustang GT parked in a parking lot. The car is positioned at an angle, showcasing its sleek design and black rims. The car's hood is black, contrasting with the green body. The Mustang GT logo is visible on the side of the car. The parking lot appears to be empty, with the car being the main focus of the video. The car's position and the absence of other vehicles suggest that the video might be a promotional or showcase video for the Mustang GT. The overall style of the video is simple and straightforward, focusing on the car and its design."
control_video = load_video("examples/training/control/wan/image_condition/validation_dataset/0.mp4")
height, width, num_frames = 480, 704, 49

# Take evenly space `num_frames` frames from the control video
indices = np.linspace(0, len(control_video) - 1, num_frames).astype(int)
control_video = [control_video[i] for i in indices]

with torch.no_grad():
    latents = pipe.prepare_latents(1, in_channels, height, width, num_frames, dtype, device, generator)
    latents_mean = torch.tensor(pipe.vae.config.latents_mean).view(1, -1, 1, 1, 1).to(latents)
    latents_std = 1.0 / torch.tensor(pipe.vae.config.latents_std).view(1, -1, 1, 1, 1).to(latents)
    control_video = pipe.video_processor.preprocess_video(control_video, height=height, width=width)
    control_video = control_video.to(device=device, dtype=dtype)
    control_latents = pipe.vae.encode(control_video).latent_dist.sample(generator=generator)
    control_latents = ((control_latents.float() - latents_mean) * latents_std).to(dtype)
    control_latents = apply_frame_conditioning_on_latents(
        control_latents,
        expected_num_frames=latents.size(2),
        channel_dim=1,
        frame_dim=2,
        frame_conditioning_type="index",
        frame_conditioning_index=0,
        concatenate_mask=False,
    )

with control_channel_concat(pipe.transformer, ["hidden_states"], [control_latents], dims=[1]):
    video = pipe(prompt, latents=latents, num_inference_steps=30, generator=generator).frames[0]

export_to_video(video, "output.mp4", fps=16)
```

<table align=center>
<tr>
<td align=center><video src="https://github.com/user-attachments/assets/0dc90217-76cf-4d77-8014-58349a1c602b">Your browser does not support the video tag.</video></td>
</tr>
</table>

### CODE REVIEW

**Comment by neph1** [comment]
fwiw, I'm trying to apply this to hunyuan video. I'll post a draft once I've got something working (or at least running)

**Comment by a-r-r-o-w** [comment]
This PR is nearly complete. Launched training runs to verify some of the different `FrameConditionType`'s on a OpenVid-1M split of 1000 videos.

Wanted to also give you a heads up @neph1 for HunyuanVideo. There were many changes made to the PR after your fork, and we'll have to try and make sure the same are made to your PR.

**Comment by neph1** [comment]
Yeah. I had not anticipated that, but I'll wait until everything is merged and then see if I can amend my pr.
